homeassistant/components/nextbus/coordinator.py

```diff
@@ -0,0 +1,79 @@
+"""NextBus data update coordinator."""
+from datetime import timedelta
+import logging
+from typing import Any
+
+from py_nextbus import NextBusClient
+from py_nextbus.client import NextBusFormatError, NextBusHTTPError, RouteStop
+
+from homeassistant.core import HomeAssistant
+from homeassistant.helpers.update_coordinator import DataUpdateCoordinator, UpdateFailed
+
+from .const import DOMAIN
+from .util import listify
+
+_LOGGER = logging.getLogger(__name__)
+
+
+class NextBusDataUpdateCoordinator(DataUpdateCoordinator):
+    """Class to manage fetching NextBus data."""
+
+    def __init__(self, hass: HomeAssistant, agency: str) -> None:
+        """Initialize a global coordinator for fetching data for a given agency."""
+        super().__init__(
+            hass,
+            _LOGGER,
+            name=DOMAIN,
+            update_interval=timedelta(seconds=30),
+        )
+        self.client = NextBusClient(output_format="json", agency=agency)
+        self._agency = agency
+        self._stop_routes: set[RouteStop] = set()
+        self._data: dict[str, Any] = {}
+
+    def add_stop_route(self, stop_tag: str, route_tag: str) -> None:
+        """Tell coordinator to start tracking a given stop and route."""
+        self._stop_routes.add(RouteStop(route_tag, stop_tag))
+
+    def remove_stop_route(self, stop_tag: str, route_tag: str) -> None:
+        """Tell coordinator to stop tracking a given stop and route."""
+        self._stop_routes.remove(RouteStop(route_tag, stop_tag))
+
+    def get_prediction_data(
+        self, stop_tag: str, route_tag: str
+    ) -> dict[str, Any] | None:
+        """Get prediction result for a given stop and route."""
+        for prediction in listify(self.data.get("predictions", [])):
+            if (
+                prediction["stopTag"] == stop_tag
+                and prediction["routeTag"] == route_tag
+            ):
+                return prediction
+
+        return None
```

Comment:
```suggestion
        return self._predictions.get(RouteStop(route_tag, stop_tag))

    def _update_predictions(self, data: dict[str, Any]) -> None:
        """Update preditions dict."""
        self._predictions.clear()
        for prediction in listify(self.data.get("predictions", [])):
            if (stop := prediction["stopTag"]) and (route := prediction["routeTag"]):
                self._predictions[RouteStop(route, stop)] = prediction
```

Performance improvement as we iterate only once over the list and not each time.